java/src/jmri/jmrix/openlcb/swing/memtool/MemoryToolPane.java

```diff
@@ -0,0 +1,427 @@
+                } else {
+                    log.error("Write failed. error code is {}", String.format("%04X", errorCode));
+                    statusField.setText("Write failed. error code is "+String.format("%016X", errorCode));
+                }
+                setRunning(false);
+                // return because we're done.
+            }
+
+            @Override
+            public void handleSuccess() {
+                log.trace("Write succeeded {} bytes", address+bytesRead);
+
+                if (cancelled) {
+                    log.debug("Cancelled");
+                    statusField.setText("Cancelled");
+                    setRunning(false);
+                    cancelled = false;
+                }
+                // next operation
+                address = address+bytesRead;
+
+                byte[] dataRead;
+                try {
+                    dataRead = getBytes();
+                    if (dataRead == null) {
+                        // end of read present
+                        setRunning(false);
+                        log.debug("Completed");
+                        statusField.setText("Completed.");
+                        inputStream.close();
+                        return;
+                    }
+                    bytesRead = dataRead.length;
+                    log.trace("write {} bytes", bytesRead);
+                } catch (IOException ex) {
+                    log.error("Error reading file",ex);
+                    return;
+                }
+                service.requestWrite(farID, space, address, dataRead, cbw);
+            }
+        };
+
+    void pushedPutButton(ActionEvent e) {
+        log.debug("Start get");
+        if (fileChooser == null) {
+            fileChooser = new JFileChooser();
+        }
+        fileChooser.setDialogTitle("Upload binary file");
+        fileChooser.rescanCurrentDirectory();
+        fileChooser.setSelectedFile(new File("memory.bin"));
+
+        int retVal = fileChooser.showOpenDialog(this);
+        if (retVal != JFileChooser.APPROVE_OPTION) { return; }
+
+        // open file and read first 64 bytes
+        File file = fileChooser.getSelectedFile();
+        log.debug("access {}", file);
+
+        byte[] dataRead;
+        try {
+            inputStream = new FileInputStream(file);
+            dataRead = getBytes();
+            if (dataRead == null) {
+                // end of read present
+                log.debug("Completed");
+                inputStream.close();
+                return;
+            }
+            bytesRead = dataRead.length;
+            log.trace("read {} bytes", bytesRead);
+        } catch (IOException ex) {
+            log.error("Error reading file",ex);
+            return;
+        }
+
+        // do first memory write
+        address = 0;
+        setRunning(true);
+        service.requestWrite(farID, space, address, dataRead, cbw);
+    }
+
+    byte[] bytes = new byte[CHUNKSIZE];
+    int bytesRead;
```

Comment:
/// How many bytes were read into the bytes[] array from the file. Used for put operation only.

Reply:
Done